Add Client-Side Gizmo

max-mrgrsk opened the pull request:
This PR introduces a client-side gizmo, addressing the limitations of the server-side gizmo and improving the overall user experience. The client-side gizmo is designed to synchronise with the client-side camera, providing a more intuitive and responsive interface. Key highlights of this PR include:

**Initial Implementation:**
Added a 80x80 canvas to the bottom right of the screen, containing a tiny three.js scene.
Created a nice @franknoirot gizmo.
Synced the gizmo's rotation with the client-side camera using quaternion values from SceneInfra.

**Enhancements:**
Implemented lerp (linear interpolation) to minimize step effects and provide fluid motion for the gizmo.
Added TailwindCSS styling for the canvas, making it rounded with a border that matches the user's theme settings.

**Future Improvements:**
Plan to add click-and-drag functionality to the gizmo for directly orienting the camera.
Fix gimbal lock over Z-axis.

**References:**
Issue #2279: Add client-side gizmo @Irev-Dev 
PR #2332: Integrate unreliable camera_drag_move
closes https://github.com/KittyCAD/modeling-app/issues/2279

**Screenshots:**
<img width="230" alt="Screenshot 2024-05-15 at 00 27 28" src="https://github.com/KittyCAD/modeling-app/assets/156543465/84c43da9-f49e-4707-8d66-1c51ab188c14">

Looking forward to your feedback and suggestions for further improvements.

max-mrgrsk committed:
draft #2279

Add client side gizmo #2279, work in progress

max-mrgrsk committed:
draft #2279

unreliableSubscriptions

max-mrgrsk committed:
draft #2279

nice Gizmo

max-mrgrsk committed:
blue ring

give the canvas a round shape and a border, wrapping rounded div element around the canvas

max-mrgrsk committed:
Refactor Gizmo Component

Extracted reusable constants
Modularized the code
Simplified the useEffect logic
Added TypeScript type annotations
Improved overall code structure and readability

franknoirot commented:
Looking good @max-mrgrsk!

You probably left these in so we could verify the syncing works as expected, but before merge we should remove the places that we create that engine-side gizmo:
https://github.com/KittyCAD/modeling-app/blob/26438270ff85e643e94265a6f1e1f071b14e82aa/src/lang/std/engineConnection.ts#L1046-L1063

https://github.com/KittyCAD/modeling-app/blob/26438270ff85e643e94265a6f1e1f071b14e82aa/src/lang/std/engineConnection.ts#L1347-L1357

I don't want to ask you to do these styling tweaks if you're trying to get this thing done, so consider it a checklist for me as a fast-follow PR if you're not able to:
- [ ] Add className "pointer-events-none" to gizmo wrapper div (for now to prevent context menu)
- [ ] Make LowerRightControls container element have these classNames: `flex flex-col items-end gap-3`
- [ ] Move gizmo into `LowerRightControls.tsx` as the first child of the section element
- [ ] Remove the fixed styling from the gizmo div so it flows in flexbox

Irev-Dev committed:
remove old gizmo

Irev-Dev committed:
fmt

Irev-Dev commented on e2e/playwright/snapshot-tests.spec.ts-snapshots/Client-side-scene-scale-should-match-engine-scale-Inch-scale-1-Google-Chrome-linux.png:
At first I was thrown by the change in the background colour of these diffs, but what think has happened is the background changed a little a while back, but such a slight change in color doesn't trigger playwright's diffing algorithm so these screenshots were never updated, but now with the gizmo change, the background colour changing is also getting updated.

Irev-Dev reviewed (commented):
This is great!

I'm surprised at how few lines of code you were able to get this working in, really great!

So here's a couple of tid-bits about our CI, and some small changes I pushed

There's an action that checks formatting that failed, easy fix was just to run `yarn fmt` and than commit and push the changes from running that.

We also have a e2e playwright tests, some of which take screenshots, because that's one of the best ways to make sure there isn't visual regressions, but because the gizmo appears on all of the screen shots, there is a lot of diffs as a result.

I went ahead and deleted the old gizmo as Frank pointed out, just because I had already checked out the branch and it was a small change.

I also quickly merge in main, just because I saw there was a conflict, it was an easy fix, but didn't want to through you into git conflicts just yet since you're new to git. We'll get there 😅 

I also looked into the gimbal lock situation, and it's not a bug on the FE, basically the camera-up that's returned from the engine is always [0,0,1], even when the camera flips upside down, I'd expect it to flip to [0,0,-1] at that point but it doesn't.

I've raised the above with the engine team, what's unfortunate about this is that I think we need to wait to merge this until that's fixed in the engine because while it's a bit of an edge case, we'd be replacing a working gizmo with one that has some quirks so we'll just have to wait sorry.

Thanks for the check list @franknoirot, those all make sense, particularly the `pointer-events-none` while it's not interact-able.

But happy with it all otherwise. I'll get onto finding you another issue!

max-mrgrsk commented:
Thanks for your feedback.
I appreciate your guidance and help. 
@franknoirot, the design you did is so nice that it's a joy to execute :D 
@Irev-Dev, thanks for all your tips. 
I'll address all the comments and updates discussed.

Irev-Dev commented:
I forgot to mention, that since our playwright action and I made commits, you'll need to pull to get these changes locally.

max-mrgrsk committed:
styling and relocation

 Add className "pointer-events-none" to gizmo wrapper div (for now to prevent context menu)
 Make LowerRightControls container element have these classNames: flex flex-col items-end gap-3
 Move gizmo into LowerRightControls.tsx as the first child of the section element
 Remove the fixed styling from the gizmo div so it flows in flexbox

max-mrgrsk commented:
> * [x]  Add className "pointer-events-none" to gizmo wrapper div (for now to prevent context menu)
> * [x]  Make LowerRightControls container element have these classNames: `flex flex-col items-end gap-3`
> * [x]  Move gizmo into `LowerRightControls.tsx` as the first child of the section element
> * [x]  Remove the fixed styling from the gizmo div so it flows in flexbox

thanks @franknoirot , all done :) looks sweet

max-mrgrsk committed:
fmt

max-mrgrsk commented:
> I also looked into the gimbal lock situation, and it's not a bug on the FE, basically the camera-up that's returned from the engine is always [0,0,1], even when the camera flips upside down, I'd expect it to flip to [0,0,-1] at that point but it doesn't.
> 
> I've raised the above with the engine team, what's unfortunate about this is that I think we need to wait to merge this until that's fixed in the engine because while it's a bit of an edge case, we'd be replacing a working gizmo with one that has some quirks so we'll just have to wait sorry.

@Irev-Dev 
Another thing is that the gizmo stops working after we exit the sketch plane. 
Shall I check on it while we are waiting?

franknoirot commented:
> Another thing is that the gizmo stops working after we exit the sketch plane.
> Shall I check on it while we are waiting?

I wonder if this issue is related to or solved by https://github.com/KittyCAD/modeling-app/pull/2397

Irev-Dev committed:
fix camera up problem

Irev-Dev commented:
Okay we're back in business with this one, video inbound, because I have a couple thoughts.

But just note because we had a bunch of conflicts with snapshots I did force push to this branch so you'll need to take some steps to get your local branch in sync.

Here's what I would do with the command line, but I know you're using the github desktop app right so I'm not sure
```bash
# We're going to delete the gizmo branch so make sure we're not on it
git checkout main 
# make sure we have an updated origin
git fetch --all
# delete our local gizmo branch
git branch -d gizmo
# checkout the origin's gizmo branch to get in sync
git checkout origin/gizmo
```

github-actions[bot] committed:
A snapshot a day keeps the bugs away! 📷🐛 (OS: ubuntu)

Irev-Dev commented on src/components/Gizmo.tsx:
```diff
@@ -42,15 +42,15 @@ export default function Gizmo() {
         camera,
         currentQuaternion,
-        clientCamera.quaternion,
+        sceneInfra.camControls.camera.quaternion,
```
this was a subtle bug where currentCamera would go stale because we swap the camera out for a orthographic camera when going into a sketch, so it would have a reference to a dead camera, currentCamera was stuck in the closure, but using the full `sceneInfra.camControls.camera.quaternion` in the animate function fixed this.

This bug only show up when entering and exiting a sketch and then the gizmo would stop updating.

Irev-Dev replied:
My apologies for missing 

> Another thing is that the gizmo stops working after we exit the sketch plane.
Shall I check on it while we are waiting?

Sorry this was the problem.

Irev-Dev commented:
https://github.com/KittyCAD/modeling-app/assets/29681384/51564fa7-a5ea-49e2-8875-3b36d8565def

Irev-Dev commented:
I'm half thinking we just merge this 🤷

Irev-Dev committed:
Merge branch 'main' into gizmo

Irev-Dev committed:
up tweak

Irev-Dev committed:
Revert "up tweak"

This reverts commit a53a0ef240ca0287440dd85a55fde9f7950b0123.

Irev-Dev committed:
test tweak

Irev-Dev committed:
tweak test

Irev-Dev committed:
Merge branch 'main' into gizmo

franknoirot commented:
I agree this can get in; I don't see that delay as devastating and I'd like to solve it with animation updates on the lossy channel or some other strategy

franknoirot committed:
Merge branch 'main' into gizmo
